utilities/joint-trajectory-player/src/Module.cpp

```diff
@@ -0,0 +1,291 @@
+    this->initializeRobotControl(parametersHandler);
+    this->instantiateSensorBridge(parametersHandler);
+
+    m_currentJointPos.resize(m_robotControl.getJointList().size());
+
+    m_qDesired.clear();
+    auto data = readStateFromFile(trajectoryFile, m_numOfJoints);
+    if (!data.first)
+    {
+        return false;
+    }
+    m_qDesired = data.second;
+
+    // check if vector is not initialized
+    if (m_qDesired.empty())
+    {
+        std::cerr << "[Module::configure] Cannot advance empty reference signals." << std::endl;
+        return false;
+    }
+
+    std::cout << "[Module::configure] Starting the experiment." << std::endl;
+
+    // Reach the first position of the desired trajectory in position control
+    m_robotControl.setReferences(m_qDesired.front(),
+                                 RobotInterface::IRobotControl::ControlMode::Position);
+
+    return true;
+}
+
+bool Module::advanceReferenceSignals()
+{
+    m_qDesired.pop_front();
+
+    // check if the vector is empty. If true the trajectory is ended
+    if (m_qDesired.empty())
+    {
+        return false;
+    }
+
+    return true;
+}
+
+bool Module::updateModule()
+{
+    bool isMotionDone;
+    bool isMimeExpired;
+    std::vector<std::pair<std::string, double>> jointlist;
+
+    m_robotControl.checkMotionDone(isMotionDone, isMimeExpired, jointlist);
+
+    if (isMotionDone)
+    {
+        if (!m_sensorBridge.advance())
+        {
+            std::cerr << "[Module::updateModule] Unable to read the sensor." << std::endl;
+            return false;
+        }
+
+        if (!m_sensorBridge.getJointPositions(m_currentJointPos))
+        {
+            std::cerr << "[Module::updateModule] Error in reading current position." << std::endl;
+            return false;
+        }
+
+        // set the reference
+        if (!m_robotControl
+                 .setReferences(m_qDesired.front(),
+                                RobotInterface::IRobotControl::ControlMode::PositionDirect))
+        {
+            std::cerr << "[Module::updateModule] Error while setting the reference position to "
+                         "iCub."
+                      << std::endl;
+            return false;
+        }
+
+        m_logJointPos.push_back(m_currentJointPos);
+
+        if (!advanceReferenceSignals())
+        {
+            std::cout << "[Module::updateModule] Experiment finished." << std::endl;
+
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool Module::close()
+{
+    std::cout << "[Module::close] I'm storing the dataset." << std::endl;
+
+    // set the file name
+    std::time_t t = std::time(nullptr);
+    std::tm tm = *std::localtime(&t);
+
+    std::ofstream stream; /**< std stream. */
+    std::stringstream fileName;
+
+    fileName << "Dataset_Measured_" << m_robotControl.getJointList().front() << "_"
+             << std::put_time(&tm, "%Y_%m_%d_%H_%M_%S") << ".txt";
+    stream.open(fileName.str().c_str());
```

Comment:
If I'm not mistaken the main purpose of this application is to collect some data to analyze them in matlab. In this case you may also use `matio-cpp` developed by @S-Dafarra. You can find a usage example in  https://github.com/dic-iit/bipedal-locomotion-framework/pull/62

Reply:
Standalone short Example usage: for reading and writing.

https://github.com/dic-iit/element_floating-base-estimation/blob/feature/contacts/code/external-cpp/FKContacts/src/LeggedOdomTest.cpp